src/dune_threaded_console/dune_threaded_console.ml

```diff
@@ -118,19 +128,30 @@ let make (module Base : S) : (module Dune_console.Backend) =
               let elapsed = now -. !last in
               if elapsed >= frame_rate then 0. else frame_rate -. elapsed
             in
-            Base.handle_user_events ~now ~time_budget mutex state
+            match Base.handle_user_events ~now ~time_budget mutex state with
+            | time -> time
+            | exception exn ->
+              let exn = Exn_with_backtrace.capture exn in
+              raise_notrace (Exn (Handle_user_events, exn))
           in
           last := new_time
         done
       with
       | Exit -> cleanup None
+      | Exn (Render, exn) -> cleanup (Some exn)
+      | Exn (Handle_user_events, exn) ->
+        (* we try to re-acquire the mutex because we can only reach this by
+           failing in [Base.handle_user_events]. and we don't know if
+           [Base.handle_user_events] released the mutex. This should be
+           impossible if the code is bug free. *)
+        (match Mutex.lock mutex with
+         | exception Sys_error _ -> ()
+         | () -> ());
+        cleanup (Some exn)
       | exn ->
```

Comment:
When can this happen now?

Reply:
Only via a programming errors. Like some of the other clauses.